Flaky test `TestSearchTagValuesV2`

`TestSearchTagValuesV2/second_batch_with_incomplete_query_-_name` sometimes fails, and it appears this may be tied to the timing of block compaction.
Example of a failing run: https://github.com/grafana/tempo/actions/runs/12831739931/job/35782895220

```
15:18:40 tempo: level=info ts=2025-01-17T15:18:40.666072638Z caller=handler.go:134 tenant=single-tenant method=GET traceID= url="/api/v2/search/tag/span.x/values?q=%7B+name%3D%22baz%22+%26%26+span.x+%3D+%7D" duration=860.231µs response_size=113 status=200
    api_test.go:594: 
        	Error Trace:	/home/runner/work/tempo/tempo/integration/e2e/api_test.go:594
        	            				/home/runner/work/tempo/tempo/integration/e2e/api_test.go:409
        	Error:      	Not equal: 
        	            	expected: []e2e.TagValue{e2e.TagValue{Type:"string", Value:"qux"}}
        	            	actual  : []e2e.TagValue{e2e.TagValue{Type:"string", Value:"bar"}, e2e.TagValue{Type:"string", Value:"qux"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,6 @@
        	            	-([]e2e.TagValue) (len=1) {
        	            	+([]e2e.TagValue) (len=2) {
        	            	+ (e2e.TagValue) {
        	            	+  Type: (string) (len=6) "string",
        	            	+  Value: (string) (len=3) "bar"
        	            	+ },
```
        
**Observations**
https://github.com/grafana/tempo/blob/main/modules/ingester/instance_search.go#L462

- Successful runs have `len(instance.completeBlocks)=0` before the test:
```
searchBlock("headBlock")
searchBlock("completingBlocks")
```
- Failed runs have `len(instance.completeBlocks)=1` before the test:
```
searchBlock("headBlock")
searchBlockWithCache("completeBlocks")
```

It seems that if block compaction triggers after the test traces are added but before querying tags, the test fails. When `searchBlockWithCache` runs, it appears to add entries from the block without filtering, which might be causing the unexpected extra result ("bar").

Example Logs
<details>
  <summary>Successful run</summary>
  
  ```
  // Testing traces are already ingested
  // The first test
  instance.SearchTagValuesV2 ->>>> req.Query=""
  instance.SearchTagValuesV2 ->>>> query="{}", req.Query=""
  instance.SearchTagValuesV2 - searchBlock ->>>> spanName="headBlock"
  instance.SearchTagValuesV2 ->>>> len(i.completeBlocks)=0
  instance.SearchTagValuesV2 - searchBlock ->>>> spanName="completingBlocks"
  instance.SearchTagValuesV2 - performSearch ->>>> query="{}", req.Query=""
  instance.SearchTagValuesV2 - performSearch ->>>> query="{}", req.Query=""
  instance.SearchTagValuesV2 - performSearch.IsEmptyQuery ->>>> query="{}", req.Query=""
  instance.SearchTagValuesV2 - performSearch.IsEmptyQuery ->>>> query="{}", req.Query=""
  // The second test
  instance.SearchTagValuesV2 ->>>> req.Query="{ name=\"baz\" && span.x = }"
  instance.SearchTagValuesV2 ->>>> query="{name=\"baz\"}", req.Query="{ name=\"baz\" && span.x = }"
  instance.SearchTagValuesV2 - searchBlock ->>>> spanName="headBlock"
  instance.SearchTagValuesV2 ->>>> len(i.completeBlocks)=0
  instance.SearchTagValuesV2 - performSearch ->>>> query="{name=\"baz\"}", req.Query="{ name=\"baz\" && span.x = }"
  instance.SearchTagValuesV2 - searchBlock ->>>> spanName="completingBlocks"
  instance.SearchTagValuesV2 - performSearch ->>>> query="{name=\"baz\"}", req.Query="{ name=\"baz\" && span.x = }"
  //
  instance.CompleteBlock ->>>> append completeBlocks, before len(i.completeBlocks) = 0
  instance.CompleteBlock ->>>> completeBlocks is updated, now len(i.completeBlocks) = 1
  ```
</details>

<details>
  <summary>Failed run</summary>
  
  ```
  // Testing traces are already ingested
  instance.CompleteBlock ->>>> append completeBlocks, before len(i.completeBlocks) = 0
  instance.CompleteBlock ->>>> completeBlocks is updated, now len(i.completeBlocks) = 1
  // The first test
  instance.SearchTagValuesV2 ->>>> req.Query=""
  instance.SearchTagValuesV2 ->>>> query="{}", req.Query=""
  instance.SearchTagValuesV2 - searchBlock ->>>> spanName="headBlock"
  instance.SearchTagValuesV2 ->>>> len(i.completeBlocks)=1
  instance.SearchTagValuesV2 - searchBlockWithCache ->>>> spanName="completeBlocks"
  instance.SearchTagValuesV2 - performSearch ->>>> query="{}", req.Query=""
  instance.SearchTagValuesV2 - performSearch.IsEmptyQuery ->>>> query="{}", req.Query=""
  instance.SearchTagValuesV2 - performSearch ->>>> query="{}", req.Query=""
  instance.SearchTagValuesV2 - performSearch.IsEmptyQuery ->>>> query="{}", req.Query=""
  LocalBlock.SearchTagValuesV2 ->>>> "x"
  // The second test
  instance.SearchTagValuesV2 ->>>> req.Query="{ name=\"baz\" && span.x = }"
  instance.SearchTagValuesV2 ->>>> query="{name=\"baz\"}", req.Query="{ name=\"baz\" && span.x = }"
  instance.SearchTagValuesV2 ->>>> len(i.completeBlocks)=1
  instance.SearchTagValuesV2 - searchBlock ->>>> spanName="headBlock"
  instance.SearchTagValuesV2 - searchBlockWithCache ->>>> spanName="completeBlocks"
  instance.SearchTagValuesV2 - performSearch ->>>> query="{name=\"baz\"}", req.Query="{ name=\"baz\" && span.x = }"
  ```
</details>

**Conclusion**
Based on these observations, it looks like the test may fail when the data is queried from a block in the `completeBlocks` list - possibly because the block compaction step happens too soon and `searchBlockWithCache` does not filter entries correctly. I might be missing some implementation details, but wanted to report this in case it indicates a bug.

Thank you for reporting this and for the detailed description.  

I can confirm that the test is flaky on my local setup as well, failing approximately 1 out of 10 times. Additionally, I observed that when `TestSearchTagValuesV2/second_batch_with_incomplete_query_-_name` fails, `TestSearchTagValuesV2/bad_query_-_unfiltered_results` also tends to fail:

```
    --- FAIL: TestSearchTagValuesV2/bad_query_-_unfiltered_results (0.00s)
        api_test.go:850: 
            	Error Trace:	/home/astoewer/Develop/tempo/integration/e2e/api_test.go:850
            	            				/home/astoewer/Develop/tempo/integration/e2e/api_test.go:595
            	            				/home/astoewer/Develop/tempo/integration/e2e/api_test.go:409
            	Error:      	"30" is not greater than "300"
            	Test:       	TestSearchTagValuesV2/bad_query_-_unfiltered_results
```

--- a/integration/e2e/api/api_test.go
+++ b/integration/e2e/api/api_test.go
@@ -26,6 +26,11 @@ const (
 	configAllInOneLocal = "../deployments/config-all-in-one-local.yaml"
 	spanX               = "span.x"
 	resourceX           = "resource.xx"
+
+	tempoPort = 3200
+
+	queryableTimeout    = 5 * time.Second        // timeout for waiting for traces to be queryable
+	queryableCheckEvery = 100 * time.Millisecond // check every 100ms for traces to be queryable
 )
 
 func TestSearchTagsV2(t *testing.T) {
@@ -59,12 +64,12 @@ func TestSearchTagsV2(t *testing.T) {
 
 	// Wait for the traces to be written to the WAL and searchable
 	require.Eventually(t, func() bool {
-		ok, err := isQueryable(tempo.Endpoint(3200))
+		ok, err := isQueryable(tempo.Endpoint(tempoPort))
 		if err != nil {
 			return false
 		}
 		return ok
-	}, 5*time.Second, 100*time.Millisecond, "traces were not queryable within timeout")
+	}, queryableTimeout, queryableCheckEvery, "traces were not queryable within timeout")
 
 	testCases := []struct {
 		name     string
@@ -307,12 +312,12 @@ func TestSearchTagValuesV2(t *testing.T) {
 
 	// Wait for the traces to be written to the WAL and searchable
 	require.Eventually(t, func() bool {
-		ok, err := isQueryable(tempo.Endpoint(3200))
+		ok, err := isQueryable(tempo.Endpoint(tempoPort))
 		if err != nil {
 			return false
 		}
 		return ok
-	}, 5*time.Second, 100*time.Millisecond, "traces were not queryable within timeout")
+	}, queryableTimeout, queryableCheckEvery, "traces were not queryable within timeout")
 
 	testCases := []struct {
 		name     string
@@ -556,7 +561,7 @@ func TestStreamingSearch_badRequest(t *testing.T) {
 	time.Sleep(time.Second * 3)
 
 	// Create gRPC client
-	c, err := util.NewSearchGRPCClient(context.Background(), tempo.Endpoint(3200))
+	c, err := util.NewSearchGRPCClient(context.Background(), tempo.Endpoint(tempoPort))
 	require.NoError(t, err)
 
 	res, err := c.Search(context.Background(), &tempopb.SearchRequest{
@@ -576,7 +581,7 @@ func callSearchTagValuesV2AndAssert(t *testing.T, svc *e2e.HTTPService, tagName,
 	urlPath := fmt.Sprintf(`/api/v2/search/tag/%s/values?q=%s`, tagName, url.QueryEscape(query))
 
 	// search for tag values
-	req, err := http.NewRequest(http.MethodGet, "http://"+svc.Endpoint(3200)+urlPath, nil)
+	req, err := http.NewRequest(http.MethodGet, "http://"+svc.Endpoint(tempoPort)+urlPath, nil)
 	require.NoError(t, err)
 
 	q := req.URL.Query()
@@ -615,7 +620,7 @@ func callSearchTagValuesV2AndAssert(t *testing.T, svc *e2e.HTTPService, tagName,
 		End:     uint32(end),
 	}
 
-	grpcClient, err := util.NewSearchGRPCClient(context.Background(), svc.Endpoint(3200))
+	grpcClient, err := util.NewSearchGRPCClient(context.Background(), svc.Endpoint(tempoPort))
 	require.NoError(t, err)
 
 	respTagsValuesV2, err := grpcClient.SearchTagValuesV2(context.Background(), grpcReq)
@@ -660,7 +665,7 @@ func callSearchTagsV2AndAssert(t *testing.T, svc *e2e.HTTPService, scope, query 
 	}
 
 	// search for tag values
-	req, err := http.NewRequest(http.MethodGet, "http://"+svc.Endpoint(3200)+urlPath, nil)
+	req, err := http.NewRequest(http.MethodGet, "http://"+svc.Endpoint(tempoPort)+urlPath, nil)
 	require.NoError(t, err)
 
 	q := req.URL.Query()
@@ -700,7 +705,7 @@ func callSearchTagsV2AndAssert(t *testing.T, svc *e2e.HTTPService, scope, query 
 		End:   uint32(end),
 	}
 
-	grpcClient, err := util.NewSearchGRPCClient(context.Background(), svc.Endpoint(3200))
+	grpcClient, err := util.NewSearchGRPCClient(context.Background(), svc.Endpoint(tempoPort))
 	require.NoError(t, err)
 
 	respTagsValuesV2, err := grpcClient.SearchTagsV2(context.Background(), grpcReq)
@@ -744,7 +749,7 @@ func prepTagsResponse(resp *searchTagsV2Response) {
 func callSearchTagsAndAssert(t *testing.T, svc *e2e.HTTPService, expected searchTagsResponse, start, end int64) {
 	urlPath := "/api/search/tags"
 	// search for tag values
-	req, err := http.NewRequest(http.MethodGet, "http://"+svc.Endpoint(3200)+urlPath, nil)
+	req, err := http.NewRequest(http.MethodGet, "http://"+svc.Endpoint(tempoPort)+urlPath, nil)
 	require.NoError(t, err)
 
 	q := req.URL.Query()
@@ -782,7 +787,7 @@ func callSearchTagsAndAssert(t *testing.T, svc *e2e.HTTPService, expected search
 		End:   uint32(end),
 	}
 
-	grpcClient, err := util.NewSearchGRPCClient(context.Background(), svc.Endpoint(3200))
+	grpcClient, err := util.NewSearchGRPCClient(context.Background(), svc.Endpoint(tempoPort))
 	require.NoError(t, err)
 
 	respTags, err := grpcClient.SearchTags(context.Background(), grpcReq)
@@ -814,7 +819,7 @@ func callSearchTagsAndAssert(t *testing.T, svc *e2e.HTTPService, expected search
 func callSearchTagValuesAndAssert(t *testing.T, svc *e2e.HTTPService, tagName string, expected searchTagValuesResponse, start, end int64) {
 	urlPath := fmt.Sprintf(`/api/search/tag/%s/values`, tagName)
 	// search for tag values
-	req, err := http.NewRequest(http.MethodGet, "http://"+svc.Endpoint(3200)+urlPath, nil)
+	req, err := http.NewRequest(http.MethodGet, "http://"+svc.Endpoint(tempoPort)+urlPath, nil)
 	require.NoError(t, err)
 
 	q := req.URL.Query()

--- a/integration/e2e/api/api_test.go
+++ b/integration/e2e/api/api_test.go
@@ -57,8 +57,14 @@ func TestSearchTagsV2(t *testing.T) {
 	batch = util.MakeThriftBatchWithSpanCountAttributeAndName(secondBatch.spanCount, secondBatch.name, secondBatch.resourceAttVal, secondBatch.spanAttVal, secondBatch.resourceAttr, secondBatch.SpanAttr)
 	require.NoError(t, jaegerClient.EmitBatch(context.Background(), batch))
 
-	// Wait for the traces to be written to the WAL
-	time.Sleep(time.Second * 3)
+	// Wait for the traces to be written to the WAL and searchable
+	require.Eventually(t, func() bool {
+		ok, err := isQueryable(tempo.Endpoint(3200))
+		if err != nil {
+			return false
+		}
+		return ok
+	}, 5*time.Second, 100*time.Millisecond, "traces were not queryable within timeout")
 
 	testCases := []struct {
 		name     string
@@ -299,8 +305,14 @@ func TestSearchTagValuesV2(t *testing.T) {
 	batch = util.MakeThriftBatchWithSpanCountAttributeAndName(secondBatch.spanCount, secondBatch.name, secondBatch.resourceAttVal, secondBatch.spanAttVal, "xx", "x")
 	require.NoError(t, jaegerClient.EmitBatch(context.Background(), batch))
 
-	// Wait for the traces to be written to the WAL
-	time.Sleep(time.Second * 3)
+	// Wait for the traces to be written to the WAL and searchable
+	require.Eventually(t, func() bool {
+		ok, err := isQueryable(tempo.Endpoint(3200))
+		if err != nil {
+			return false
+		}
+		return ok
+	}, 5*time.Second, 100*time.Millisecond, "traces were not queryable within timeout")
 
 	testCases := []struct {
 		name     string
@@ -866,6 +878,26 @@ func lenWithoutIntrinsic(resp searchTagsV2Response) int {
 		size += len(scope.Tags)
 	}
 	return size
+}
+
+// isQueryable returns true if the data is queryable and not empty
+func isQueryable(host string) (bool, error) {
+	resp, err := http.Get(fmt.Sprintf("http://%s/api/search/tag/service.name/values", host))
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return false, fmt.Errorf("status code is not ok: %d", resp.StatusCode)
+	}
+
+	var searchResp searchTagValuesResponse
+	if err := json.NewDecoder(resp.Body).Decode(&searchResp); err != nil {
+		return false, err
+	}
+
+	return len(searchResp.TagValues) > 0, nil
 }
 
 type ScopedTags struct {

--- a/integration/e2e/api/compression_test.go
+++ b/integration/e2e/api/compression_test.go
@@ -37,9 +37,9 @@ func TestCompression(t *testing.T) {
 	info := tempoUtil.NewTraceInfo(time.Now(), "")
 	require.NoError(t, info.EmitAllBatches(c))
 
-	apiClient := httpclient.New("http://"+tempo.Endpoint(3200), "")
+	apiClient := httpclient.New("http://"+tempo.Endpoint(tempoPort), "")
 
-	apiClientWithCompression := httpclient.NewWithCompression("http://"+tempo.Endpoint(3200), "")
+	apiClientWithCompression := httpclient.NewWithCompression("http://"+tempo.Endpoint(tempoPort), "")
 
 	util.QueryAndAssertTrace(t, apiClient, info)
 	queryAndAssertTraceCompression(t, apiClientWithCompression, info)

--- a/integration/e2e/api/cross_cluster_reads_test.go
+++ b/integration/e2e/api/cross_cluster_reads_test.go
@@ -47,7 +47,7 @@ func TestCrossClusterReads(t *testing.T) {
 	require.NoError(t, tempoDistributorA.WaitSumMetrics(e2e.Equals(util.SpanCount(expected)), "tempo_distributor_spans_received_total"))
 
 	// read from cluster B
-	apiClient := httpclient.New("http://"+tempoQueryFrontendB.Endpoint(3200), "")
+	apiClient := httpclient.New("http://"+tempoQueryFrontendB.Endpoint(tempoPort), "")
 
 	// query an in-memory trace
 	util.QueryAndAssertTrace(t, apiClient, info)

--- a/integration/e2e/api/https_test.go
+++ b/integration/e2e/api/https_test.go
@@ -59,14 +59,14 @@ func TestHTTPS(t *testing.T) {
 	info := tempoUtil.NewTraceInfo(time.Now(), "")
 	require.NoError(t, info.EmitAllBatches(c))
 
-	apiClient := httpclient.New("https://"+tempo.Endpoint(3200), "")
+	apiClient := httpclient.New("https://"+tempo.Endpoint(tempoPort), "")
 
 	// trust bad certs
 	defaultTransport := http.DefaultTransport.(*http.Transport).Clone()
 	defaultTransport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
 	apiClient.WithTransport(defaultTransport)
 
-	echoReq, err := http.NewRequest("GET", "https://"+tempo.Endpoint(3200)+"/api/echo", nil)
+	echoReq, err := http.NewRequest("GET", "https://"+tempo.Endpoint(tempoPort)+"/api/echo", nil)
 	require.NoError(t, err)
 	resp, err := apiClient.Do(echoReq)
 	require.NoError(t, err)
@@ -78,7 +78,7 @@ func TestHTTPS(t *testing.T) {
 	util.SearchTraceQLAndAssertTrace(t, apiClient, info)
 
 	creds := credentials.NewTLS(&tls.Config{InsecureSkipVerify: true})
-	grpcClient, err := util.NewSearchGRPCClientWithCredentials(context.Background(), tempo.Endpoint(3200), creds)
+	grpcClient, err := util.NewSearchGRPCClientWithCredentials(context.Background(), tempo.Endpoint(tempoPort), creds)
 	require.NoError(t, err)
 
 	now := time.Now()

--- a/integration/e2e/api/multi_tenant_test.go
+++ b/integration/e2e/api/multi_tenant_test.go
@@ -101,10 +101,10 @@ func testSearch(t *testing.T, tenant string, tenantSize int) {
 	time.Sleep(time.Second * 3)
 
 	// test echo
-	util.AssertEcho(t, "http://"+tempo.Endpoint(3200)+"/api/echo")
+	util.AssertEcho(t, "http://"+tempo.Endpoint(tempoPort)+"/api/echo")
 
 	// client will have testcase tenant id
-	apiClient := httpclient.New("http://"+tempo.Endpoint(3200), tenant)
+	apiClient := httpclient.New("http://"+tempo.Endpoint(tempoPort), tenant)
 
 	// check trace by id
 	resp, err := apiClient.QueryTrace(info.HexID())
@@ -177,7 +177,7 @@ func testSearch(t *testing.T, tenant string, tenantSize int) {
 	grpcCtx, err = user.InjectIntoGRPCRequest(grpcCtx)
 	require.NoError(t, err)
 
-	grpcClient, err := util.NewSearchGRPCClient(grpcCtx, tempo.Endpoint(3200))
+	grpcClient, err := util.NewSearchGRPCClient(grpcCtx, tempo.Endpoint(tempoPort))
 	require.NoError(t, err)
 
 	time.Sleep(2 * time.Second) // ensure that blocklist poller has built the blocklist

--- a/integration/e2e/api/query_range_test.go
+++ b/integration/e2e/api/query_range_test.go
@@ -57,11 +57,11 @@ sendLoop:
 		"{} | compare({status=error})",
 	} {
 		t.Run(query, func(t *testing.T) {
-			callQueryRange(t, tempo.Endpoint(3200), query, debugMode)
+			callQueryRange(t, tempo.Endpoint(tempoPort), query, debugMode)
 		})
 	}
 
-	res := doRequest(t, tempo.Endpoint(3200), "{. a}")
+	res := doRequest(t, tempo.Endpoint(tempoPort), "{. a}")
 	require.Equal(t, 400, res.StatusCode)
 }
 

